stm32: added microseconds/ticks conversions and pulse_width_us handling.

vlad-g-maglasu opened the pull request:
<!-- Thanks for submitting a Pull Request! We appreciate you spending the
     time to improve MicroPython. Please provide enough information so that
     others can review your Pull Request.

     Before submitting, please read:
     https://github.com/micropython/micropython/blob/master/CODEOFCONDUCT.md
     https://github.com/micropython/micropython/wiki/ContributorGuidelines

     Please check any CI failures that appear after your Pull Request is opened.
-->

### Summary

<!-- Explain the reason for making this change. What problem does the pull request
     solve, or what improvement does it add? Add links if relevant. -->

Added support for pulse_width_us and fixed unsafe conversions in the STM32 Timer code. Previously there was no direct API to get/set pulse width in microseconds and the code performed an unsafe cast from an mp_obj_t pointer to an integer. This change adds a safe µs↔ticks conversion, exposes pulse_width_us() correctly to Python (returns an mp_obj_t), and accepts pulse_width_us as an optional channel init parameter to set the initial compare value.

### Testing

<!-- Explain what testing you did, and on which boards/ports. If there are
     boards or ports that you couldn't test, please mention this here as well.

     If you leave this empty then your Pull Request may be closed. -->

Hardware

Board: ARDUINO PORTENTA H7
Setup: flashed the build produced from this branch to the board. Connected a standard hobby servo to the PWM pin (same wiring used previously).
Runtime test: used UDP control loop (my test script in main.py) to send steering/speed values to the board and verified the servo responded to pwm_channel_X.pulse_width_us(...) updates.

- Reproduction steps:
Flash the build for ARDUINO_PORTENTA_H7.
Run the main.py on the board (or your equivalent UDP control program).
Send a UDP message like: echo '{"mechanic.engine.steering": -1}' | nc -u <BOARD_IP> 'password' and observe servo moves to expected position. Repeat with other steering values.

Software tests:
Ran basic local sanity checks (file-level static checks in editor). CI builds will check full compilation across targets; please review CI results if any failures appear.

### Trade-offs and Alternatives

<!-- If the Pull Request has some negative impact (i.e. increased code size)
     then please explain why you think the trade-off improvement is worth it.
     If you can think of alternative ways to do this, please explain that here too.

Delete this heading if not relevant (i.e. small fixes) -->

- Backwards compatibility: the new pulse_width_us constructor keyword is additive; it sets an initial compare value only.
- Runtime API (pwm_channel.pulse_width_us(...)) is unchanged. Internal helper signature changed (now returning mp_obj_t) but that helper is internal to the stm32 port.
- Size/perf: negligible. Only a small constant (US_PER_SEC) and a tiny conversion change; no significant code-size impact expected.

Notes: conversions do not account for center-aligned timer counting (code uses simple up-counting conversion). This matches the decision to keep conversion straightforward and consistent across timer modes. If align-aware behavior is desired later, we can add an optional parameter or a separate helper.

dpgeorge commented on ports/stm32/timer.c:
```diff
@@ -1148,6 +1165,7 @@ static mp_obj_t pyb_timer_channel(size_t n_args, const mp_obj_t *pos_args, mp_ma
         { MP_QSTR_pin,                 MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE} },
         { MP_QSTR_pulse_width,         MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0} },
         { MP_QSTR_pulse_width_percent, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE} },
+        { MP_QSTR_pulse_width_us,      MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE} },
```
Adding this new argument here will change the `args[...]` index for the two arguments below (compare, polarity).  So please adjust their indices as appropriate.

vlad-g-maglasu replied:
OC uses args[7] (compare) and args[8] (polarity), IC uses args[8] (polarity), ENC uses args[8] (polarity) — all properly shifted to account for the two new entries at indices 5 and 6.

dpgeorge commented on ports/stm32/timer.c:
```diff
@@ -1628,6 +1649,30 @@ static mp_obj_t pyb_timer_channel_pulse_width_percent(size_t n_args, const mp_ob
+/// up-counting; this keeps the conversion straightforward and consistent
+/// across timer modes. The value may be an integer or a
+/// floating-point number for greater precision when floating point support
+/// is enabled. For example, passing 1000 sets the pulse width to 1 ms.
```
Please move this description into the docs: `docs/library/pyb.Timer.rst`.

vlad-g-maglasu replied:
Moved the descriptions into the docs: docs/library/pyb.Timer.rst.

dpgeorge reviewed (commented):
Thanks for the contribution, this is a nice addition.

Instead of microseconds it could use nanoseconds (eg `pulse_width_ns`) which would allow slightly better resolution, especially at small pulse width.  That would also match `machine.PWM.duty_ns()`.  Anyway, that's just a suggestion.  Happy to keep it as microseconds if you think that's best.

dpgeorge commented:
> and fixed unsafe conversions in the STM32 Timer code

I don't see any fix for this, what exactly is the unsafe conversion?

vlad-g-maglasu commented:
Hello, sorry for my late response. What I meant with "Unsafe conversion". Doing a direct C-style cast like (uint32_t)us could be unsafe because it evaluates the raw memory address of the pointer, resulting in complete garbage data for your timer. Using MicroPython's C API, like mp_obj_get_int(us), will safely convert the Python integer object into a C integer. For our project we needed the conversion in microseconds. But I can add a conversion in nanoseconds as well. I will move the decription into the docs, and I will update the indexes for the argumets.

dpgeorge commented:
OK, thanks for the explanation.

I look forward to the updated PR.

Commit:
stm32/timer: Add pulse_width_us and pulse_width_ns to TimerChannel.

This allows setting the pulse width in microseconds and nanoseconds, which
is useful for certain applications.

Signed-off-by: vlad.maglasu <vlad.maglasu@vampirebyte.ro>

dpgeorge commented:
@vlad-g-maglasu thanks for updating the PR.

I see you added the nanosecond feature alongside the microsecond feature.  But I think we only need one of these.  I think having just nanoseconds is the best option because it's more precise than microseconds, and makes sure the firmware size is not too big.

Does having just nanoseconds work for your application?

vlad-g-maglasu commented:
Thanks for the feedback! Unfortunately, dropping the microsecond feature won't work for our use case. Our project relies on microseconds because the motors we're using simply aren't precise enough to benefit from nanosecond resolution.

If keeping the firmware size down is the main priority, I'm happy to remove the nanosecond feature and just keep microseconds. Otherwise, I think we should keep both to support different hardware needs. Let me know what you think is best.

dpgeorge commented:
> Unfortunately, dropping the microsecond feature won't work for our use case. Our project relies on microseconds because the motors we're using simply aren't precise enough to benefit from nanosecond resolution.

But can't you just use `tim.pulse_width_ns(microsecond * 1000)` instead of `tim.pulse_width_us(microsecond)`?  That should be equivalent.

vlad-g-maglasu commented:
That is mathematically equivalent, but I think we should keep the microsecond API for domain accuracy and readability.
Most of the standard RC servos and ESCs are documented and operate natively on microsecond pulse widths (typically the 1000us - 2000us range). Our entire configuration is built around this standard, with constants like ESC_MIN_US = 1080.0 and STEERING_MIN_US = 1200. Abstracting this up into millions of nanoseconds via runtime multiplication (* 1000) obscures the intent of the code and misaligns it with the actual hardware data sheets. It’s much cleaner to use the API that matches the hardware's native units.

dpgeorge commented:
> It’s much cleaner to use the API that matches the hardware's native units.

Yes I agree with that, it's always best to use natural units where possible.

OK, so let's leave both us and ns in.

dpgeorge committed:
stm32/boards/ESPRUINO_PICO: Disable MICROPY_PY_PYB_LEGACY.

To reduce firmware size, so it fits with recent additions to `pyb.Timer`.

Signed-off-by: Damien George <damien@micropython.org>
